File: app/src/main/java/com/cn29/aac/binding/SpinnerBindingUtil.java
package com.cn29.aac.binding;

import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import androidx.appcompat.widget.AppCompatSpinner;
import androidx.databinding.BindingAdapter;
import androidx.databinding.InverseBindingAdapter;
import androidx.databinding.InverseBindingListener;

/**
 * Created by Charles Ng on 7/9/2017.
 */

public class SpinnerBindingUtil {

  @BindingAdapter(value = {"selectedValue", "selectedValueAttrChanged"}, requireAll = false)
  public static void bindSpinnerData(AppCompatSpinner pAppCompatSpinner, String newSelectedValue, final InverseBindingListener newTextAttrChanged) {
    pAppCompatSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
      @Override
      public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        newTextAttrChanged.onChange();
      }
      @Override
      public void onNothingSelected(AdapterView<?> parent) {
      }
    });
    if (newSelectedValue != null) {
      int pos = ((ArrayAdapter<String>) pAppCompatSpinner.getAdapter()).getPosition(newSelectedValue);
      pAppCompatSpinner.setSelection(pos, true);
    }
  }

  @InverseBindingAdapter(attribute = "selectedValue", event = "selectedValueAttrChanged")
  public static String captureSelectedValue(AppCompatSpinner pAppCompatSpinner) {
    return (String) pAppCompatSpinner.getSelectedItem();
  }

}

File: app/src/main/java/com/cn29/aac/binding/SpinnerBindingUtil.kt
package com.cn29.aac.binding

import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import androidx.appcompat.widget.AppCompatSpinner
import androidx.databinding.BindingAdapter
import androidx.databinding.InverseBindingAdapter
import androidx.databinding.InverseBindingListener

/**
 * Created by Charles Ng on 7/9/2017.
 */
object SpinnerBindingUtil {
    @BindingAdapter(value = ["selectedValue", "selectedValueAttrChanged"],
                    requireAll = false)
    fun bindSpinnerData(pAppCompatSpinner: AppCompatSpinner,
                        newSelectedValue: String?,
                        newTextAttrChanged: InverseBindingListener) {
        pAppCompatSpinner.onItemSelectedListener = object :
                OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?,
                                        view: View,
                                        position: Int,
                                        id: Long) {
                newTextAttrChanged.onChange()
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
        if (newSelectedValue != null) {
            val pos = (pAppCompatSpinner.adapter as ArrayAdapter<String?>).getPosition(
                    newSelectedValue)
            pAppCompatSpinner.setSelection(pos, true)
        }
    }

    @InverseBindingAdapter(attribute = "selectedValue",
                           event = "selectedValueAttrChanged")
    fun captureSelectedValue(pAppCompatSpinner: AppCompatSpinner): String {
        return pAppCompatSpinner.selectedItem as String
    }
}